src/utils/utils.ts

```diff
@@ -62,53 +64,177 @@ function standardErrorReplacer(key: string, value: any) {
+      maxColumnLengths[column] = Math.max(
+        maxColumnLengths[column] || 0,
+        cellLength, // Use the length of the encoded value
+      );
+    }
+  }
+
+  // After this point, maxColumnLengths will have been filled with all the necessary keys.
+  // Thus, the column keys can be derived from it.
+  const columns = Object.keys(maxColumnLengths);
+  // If headers are provided, add them to your output first
+  if (optionColumns != null) {
+    for (let i = 0; i < optionColumns.length; i++) {
+      const column = optionColumns[i];
+      const maxColumnLength = maxColumnLengths[column];
+      // Options.headers is definitely defined as optionHeaders != null
+      if (!Array.isArray(options!.columns)) {
+        options!.columns![column] = maxColumnLength;
+      }
+      if (includeHeaders) {
+        output += column.padEnd(maxColumnLength);
+        if (i !== optionColumns.length - 1) {
+          output += '\t';
+        } else {
+          output += '\n';
+        }
+      }
+    }
+  }
+
+  for (const row of rows) {
+    let formattedRow = '';
+    if (options.includeRowCount) {
+      formattedRow += `${++rowCount}\t`;
+    }
+    for (const column of columns) {
+      // Assume row[key] has been already encoded as a string or null
+      const cellValue =
+        row[column] == null || row[column].length === 0 ? 'N/A' : row[column];
+      formattedRow += `${cellValue.padEnd(maxColumnLengths[column] || 0)}\t`;
+    }
+    output += formattedRow.trimEnd() + '\n';
+  }
+
+  return output;
+}
+
+/**
+ * Formats a message suitable for output.
+ * @param msg - The msg that needs to be formatted.
+ * @see {@link outputTableFormatter} for information regarding usage where `msg.type === 'table'`.
+ * @returns
+ */
 function outputFormatter(msg: OutputObject): string | Uint8Array {
   let output = '';
   if (msg.type === 'raw') {
     return msg.data;
   } else if (msg.type === 'list') {
-    for (let elem in msg.data) {
-      // Empty string for null or undefined values
-      if (elem == null) {
-        elem = '';
-      }
-      output += `${msg.data[elem]}\n`;
+    for (const elem of msg.data) {
+      // Convert null or undefined to empty string
+      output += `${elem != null ? encodeNonPrintable(elem) : ''}\n`;
     }
   } else if (msg.type === 'table') {
-    for (const key in msg.data[0]) {
-      output += `${key}\t`;
-    }
-    output = output.substring(0, output.length - 1) + `\n`;
-    for (const row of msg.data) {
-      for (const key in row) {
-        let value = row[key];
-        // Empty string for null or undefined values
-        if (value == null) {
-          value = '';
-        }
-        value = value.toString();
-        // Remove the last line terminator if it exists
-        // This may exist if the value is multi-line string
-        value = value.replace(/(?:\r\n|\n)$/, '');
-        output += `${value}\t`;
+    return outputTableFormatter(msg.data, msg.options);
```

Comment:
If this is going to be a separate function, then factor out all the other ones too.

And I suggest:

* `outputFormatterTable`
* `outputFormatterList`
* `outputFormatterDict` ...

Reply:
resolved by 4fd8ce70aad2d93348645a602d19d795595e45eb